### PR TITLE
A fix for transaction malleability

### PR DESCRIPTION
**Pull request by bitcartel**
This PR fixes transaction malleability by not including the sigscript of transaction inputs and joinsplit sigs when hashing the txid.

This PR supercedes PR #1101 which was a minimal solution based on a new serialization flag.

This PR introduces GetTxid() to distinguish between getting a transaction id and the double sha256 hash.

The key changes are:
- Adding GetTxid() method to CTransaction which makes a copy of the transaction, clearing out the sigscript and joinsplitsig fields, before hashing.
- Verifying that every call to GetHash() actually wants a txid, and replacing with GetTxid(). 
- Renaming GetHash() to GetSerializeHash()
  - Rationale: In future, upstream code we want to merge will use GetHash() but we don't know the intent.  We should check to see if the intent is to receive a txid (most likely) in which case we replace with GetTxid(), or if upstream actually wants a double hash of the transaction we can use GetSerializeHash().
- Updated genesis data in chainparams.cpp

Note that coinbase transactions are excluded as they need the sigscript hashed to help avoid duplicate txids per BIP34:
  - This modification is related to a question from @ebfull on PR #1101 - "Can we think of a way this change allows us to construct two transactions with the same txid which can simultaneously appear in the blockchain? My guess is it would be possible to construct a coinbase transaction of such a form... this surely breaks invariants."

This PR Passes all tests in test_bitcoin (test data was updated in bloom_tests, miner_tests and script_tests).

### CODE REVIEW

**Comment by ebfull** [comment]
Are we preventing coinbases from appearing with the same txid *as a consensus rule*? The changes in this PR sound like they only prevent it locally, but a miner might be able to abuse it to cause double-spends or something.

**Comment by ebfull** [comment]
See BIP 34 and http://bitcoin.stackexchange.com/questions/30764/why-block-height-is-required-in-coinbase

**Comment by bitcartel** [comment]
Only prevents it locally.  Not consensus.

**Comment by bitcartel** [comment]
Thx. Based on BIP34 I think we just need to have the txid of a coinbase tx be computed the old way, to include the sigscript.  So I can then remove the change in CreateNewBlock.  Will keep you posted.

**Comment by defuse** [comment]
It might be worthwhile to add domain separation between `GetSerializeHash()` and `GetTxid()`. For example, prefix the serialization with a string like "TXID" or something (just make sure it isn't a valid serialization). I can't think of a good reason this would be necessary, but here's a super contrived protocol that would require this domain separation:

There's a server that holds a secret random transaction. Everyone on the Internet is allowed to do three things: 1. Query for the secret transaction's `GetSerializeHash()`, 2. Tell the server to re-randomize the transaction, and 3. Submit a guess for the txid. If you guess the txid correctly, the server sends you 100 bitcoins. Without domain separation, this game can be won by re-randomizing and submitting the `GetSerializeHash()` as the txid, hoping to land on one of the transactions that have the same txid as their hash. With domain separation, it shouldn't be winnable.

We *don't* need domain separation if nowhere in the Zcash protocol are there interactions between the two like this, but I think it's easier to add domain separation than to check all the uses.

(Background on chosen protocol attacks: https://www.schneier.com/academic/archives/1997/01)/protocol_interaction.html)

**Comment by bitcartel** [comment]
BIP62 provides background on tx malleability: https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki

**Comment by daira** [comment]
+1 for domain separation here. Note that we  have a domain separation scheme for SHA-256 hashes already; 0xB1 is the next unused lead byte. (I don't know whether that forms an invalid first byte for a serialised transaction.)

**Comment by daira** [comment]
We could use 0xB1 for serialised transactions and 0xB2 for txids, potentially. Would that break any valuable Bitcoin compatibility properties?

**Comment by daira** [comment]
Alternatively, we could use BLAKE2b-256 with a personalisation string.

**Comment by daira** [comment]
Re: 92f7426, that's going in the wrong direction: https://github.com/zcash/zcash/issues/967 . All the fields *have* to be made non-const to avoid undefined behaviour.

**Comment by bitcartel** [comment]
@daira Yes the const_casts should be cleaned up.  Let's plan to do #967 after the z8 release.

**Comment by bitcartel** [comment]
At the moment there are zero calls to GetSerializeHash() aka GetHash() because all calls had the intention to receive a transaction id, so they were replaced with GetTxid().

I get the gist of what has been written above regarding chosen protocol attacks.  I'll defer to others as to whether or not to proceed with domain separation.

I would say though that the ecosysytem need to update their code to verify non-malleable txids (by skipping over sigscript and joinsplit sigs like in this PR).  They already make use of sha256 crypto code/libraries so this is fairly simple and familiar to them.  If we adopt domain separation, as described above, they would also have to introduce blake2 crypto into their project (since Bitcoin does not use it in any way).

Regarding implementation, what should the input to the blake2b function be?  Can it be the 32 byte txid (the result of double sha256 of tx skipping over sigs)  with the personalization string being a fixed set of data e.g. "ZCASH...." or something from the transaction itself, such as first 16 bytes of the first tx's output address?

**Comment by zookozcash** [comment]
I'm +0 on domain separation, because I am in favor of domain separation in principle but I don't know all the details of this. I'd suggest Daira be the one to decide if we do it or not.

Likewise I'm +0 on BLAKE2 for the same reasons. :-) I think our ecosystem partners will mostly just be deploying code that we wrote, and if they want to verify txids themselves, it is easy to find a BLAKE2 package in their favorite programming language (https://blake2.net/#sw) or it is already included in their favorite crypto library (https://en.wikipedia.org/wiki/Comparison_of_cryptography_libraries#Hash_functions).

So, likewise I think Daira should decide SHA256 or BLAKE2 for this.

**Comment by daira** [comment]
I will look at the PR and the existing crypto in more detail today and make a decision on the domain separation.

File: src/primitives/transaction.cpp [hunk 1]
```diff
@@ -223,3 +226,33 @@ std::string CTransaction::ToString() const
         str += "    " + vout[i].ToString() + "\n";
     return str;
 }
+
+// Return a txid which is non-malleable.
```
Comment (daira) [inline src/primitives/transaction.cpp]: This doesn't return the txid, it only sets the variable.

**Comment by str4d** [comment]
Note that the reason I kept the use of double-SHA256 for the PoW difficulty check was because the block header hash was used in many places, and thus it would have been more invasive a change than we wanted at the time.

If I understand this PR correctly, `GetHash()` is now only used by the miner? If that's the case, we could consider changing the difficulty check to BLAKE2b, so that the miner only depends on one hash algorithm. (Generation of a block header template would still implicitly require SHA256, but a standalone miner like proposed in #1156 would simply be fetching the header template.)

**Comment by daira** [comment]
If ``GetSerializeHash`` is indeed only used by the miner then my inclination is to use BLAKE2b with personalisation, since in that case we will be changing the semantics of all of the hashes (due to nonmalleability for txids and due to Equihash+header changes for mining), therefore there is nothing to potentially remain compatible.

**Comment by daira** [comment]
Note that single BLAKE2b-256 suffices to replace *double* SHA-256, because the only motivation for doubling IIUC is to avoid length extension attacks, and BLAKE2b's finalisation flag already prevents those.

**Comment by daira** [comment]
A remaining question is whether to change other uses of ``SerializeHash``, e.g. in ``CTxOut::GetHash()``. My inclination is yes.

**Comment by bitcartel** [comment]
@str4d The block header class has its own GetHash() method which computes a double sha256.  Nothing has changed there. Just wanted to clarify in case there's any confusion.  Every single transaction.GetHash() call has been replaced with tx.GetTxid().

@daira The miner class used to call GetHash() on transaction objects. Those calls have been replaced with GetTxid().  I manually checked them and I believe they all intend to receive a transaction id, and not a double sha256 (worth eyeballing in case I erred!).  Thus, GetSerializeHash() is never called by the miner or anywhere else.

**Comment by daira** [comment]
On reflection, I don't think domain separation is necessary. I'll need to check all the calls to ``GetTxid``, but assuming that they are all intending to get a txid, there should be no vulnerability to chosen protocol attacks there.

We should remove the ``GetSerializeHash`` method.

**Comment by bitcartel** [comment]
Ok, so let's remove `GetSerializeHash` from CTransaction (and CMutableTransaction).

We should keep the `UpdateHash` method as the `hash` member variable is used by overloaded operators == and !=.

We could choose to remove the `GetHash` method from CTxOut since nobody calls it or keep things simple and just leave it for now.

**Comment by bitcartel** [comment]
Self-ACK...

**Comment by ebfull** [comment]
This PR is a bit scary for me. Not because it isn't done well, but because there are almost certainly side-effects we aren't considering. I can't think of any, and my previous nits have been fixed, so I am tempted to ACK this. I'll take a look at it again tonight when I get home.

**Comment by bitcartel** [comment]
@ebfull Possible but unlikely scenario for transparent txs.  On the last day of every month, Employer (address A) pays salary to Worker (address B) who transfers a fixed amount to Savings (address C).  Someone could spot the pattern and craft a transaction, using the UTXO at B as an input, and an output to C.  This might result in a tx with the same txid as the legitimate tx which the Worker has not yet created.  Broadcasting this txid with fake sigs will result in the txid being rejected from entering the mempool and the txid getting placed in the recentRejects filter, so that eventually when the legitimate tx is broadcast it is rejected too.  However, this scenario is unlikely since the change address should be different each time, no guarantee a wallet will use that UTXO and not a bunch of others, output order could differ, nLockTime (timestamp) might be set, etc. making it hard to predict and craft a tx with the same txid.

**Comment by ebfull** [comment]
Considering this kind of stuff five days before an audit is really suboptimal, but given my understanding of the PR and Bitcoin, **ACK**.

File: src/primitives/transaction.cpp [hunk 2]
```diff
@@ -223,3 +221,32 @@ std::string CTransaction::ToString() const
...
+        // Clear sigscript from all transaction inputs.
+        for (CTxIn & txIn : tx.vin) {
+            txIn.scriptSig.clear();
+        }
+
+        // Clear joinSplitSig by filling the buffer with zero
+        tx.joinSplitSig.assign(0);
+    }
+
+    *const_cast<uint256*>(&txid) = SerializeHash(tx);
```
Comment (str4d) [inline src/primitives/transaction.cpp]: Does this mean that we aren't actually "skipping" the sigs as such, but instead are including them as zeroes? That needs to be clear for people to recreate the txids.

File: src/test/bloom_tests.cpp [hunk 3]
```diff
@@ -254,10 +259,10 @@ BOOST_AUTO_TEST_CASE(merkle_block_2)
     for (unsigned int i = 0; i < vMatched.size(); i++)
         BOOST_CHECK(vMatched[i] == merkleBlock.vMatchedTxn[i].second);
 
-    // Match an output from the second transaction (the pubkey for address 1DZTzaBHUDM7T3QvUKBz4qXMRpkg8jsfB5)
-    // This should match the third transaction because it spends the output matched
-    // It also matches the fourth transaction, which spends to the pubkey again
-    filter.insert(ParseHex("044a656f065871a353f216ca26cef8dde2f03e8c16202d2e8ad769f02032cb86a5eb5e56842e92e19141d60a01928f8dd2c875a390f67c1f6c94cfc617c0ea45af"));
+    // Match an output from the second transaction (the pubkey for address mjuZa8Dy12HKyUNjc1whNTxRaGU4gGGLxY)
+    // It also matches the third transaction, which spends to the pubkey again
+    // This should match the last transaction because it spends the output matched
```
Comment (str4d) [inline src/test/bloom_tests.cpp]: Which code changes cause this semantic change in the test?

File: src/test/bloom_tests.cpp [hunk 4]
```diff
@@ -308,10 +313,10 @@ BOOST_AUTO_TEST_CASE(merkle_block_2_with_update_none)
     for (unsigned int i = 0; i < vMatched.size(); i++)
         BOOST_CHECK(vMatched[i] == merkleBlock.vMatchedTxn[i].second);
 
-    // Match an output from the second transaction (the pubkey for address 1DZTzaBHUDM7T3QvUKBz4qXMRpkg8jsfB5)
-    // This should not match the third transaction though it spends the output matched
-    // It will match the fourth transaction, which has another pay-to-pubkey output to the same address
-    filter.insert(ParseHex("044a656f065871a353f216ca26cef8dde2f03e8c16202d2e8ad769f02032cb86a5eb5e56842e92e19141d60a01928f8dd2c875a390f67c1f6c94cfc617c0ea45af"));
+    // Match an output from the second transaction (the pubkey for address mjuZa8Dy12HKyUNjc1whNTxRaGU4gGGLxY)
+    // It will match the third transaction, which has another pay-to-pubkey output to the same address
+    // This should not match the last transaction though it spends the output matched
```
Comment (str4d) [inline src/test/bloom_tests.cpp]: Ditto.

**Comment by str4d** [comment]
Impl ACK modulo comments. I am not enough of a Bitcoiner to understand all the possible side-effects of not including the signatures, but I'll assume that they are all less bad than the known issue of transaction malleability.

File: src/test/bloom_tests.cpp (hunk 4 above)
Comment (bitcartel) [inline src/test/bloom_tests.cpp]: When deserializing the hex string into a block object, the txids are computed the new way so they don't match the txids used in the test.

Comment (str4d) [inline src/test/bloom_tests.cpp]: I understand that; I was asking why the semantic ordering in the comment changed. Are the transactions ordered by txid?

Comment (bitcartel) [inline src/test/bloom_tests.cpp]: I see what you mean.   I generated the transactions by hand to try and match the original test data (transactions in a real block).  The exact order of the transactions was based on the regtest miner.

File: src/primitives/transaction.cpp (hunk 2 above)
Comment (bitcartel) [inline src/primitives/transaction.cpp]: Ok, will make a ticket about documenting this with examples.

**Comment by bitcartel** [comment]
@zkbot r+

**Pull request merged**